EKF/ekf_helper.cpp

```diff
@@ -1819,3 +1816,137 @@ void Ekf::stopFlowFusion()
+
+		// update transformation matrix from body to world frame using the current estimate
+		_R_to_earth = Dcmf(_state.quat_nominal);
+
+		// update the rotation matrix using the new yaw value
+		// determine if a 321 or 312 Euler sequence is best
+		if (fabsf(_R_to_earth(2, 0)) < fabsf(_R_to_earth(2, 1))) {
+			// use a 321 sequence
+			Eulerf euler321(_state.quat_nominal);
+			euler321(2) = new_yaw;
+			_R_to_earth = euler321;
+
+		} else {
+			// Calculate the 312 Tait-Bryan rotation sequence that rotates from earth to body frame
+			// We use a 312 sequence as an alternate when there is more pitch tilt than roll tilt
+			// to avoid gimbal lock
+			Vector3f rot312;
+			rot312(0) = new_yaw;
+			rot312(1) = asinf(_R_to_earth(2, 1));
+			rot312(2) = atan2f(-_R_to_earth(2, 0), _R_to_earth(2, 2));
+
+			float c2 = cosf(rot312(2));
+			float s2 = sinf(rot312(2));
+			float s1 = sinf(rot312(1));
+			float c1 = cosf(rot312(1));
+			float s0 = sinf(rot312(0));
+			float c0 = cosf(rot312(0));
+
+			_R_to_earth(0, 0) = c0 * c2 - s0 * s1 * s2;
+			_R_to_earth(1, 1) = c0 * c1;
+			_R_to_earth(2, 2) = c2 * c1;
+			_R_to_earth(0, 1) = -c1 * s0;
+			_R_to_earth(0, 2) = s2 * c0 + c2 * s1 * s0;
+			_R_to_earth(1, 0) = c2 * s0 + s2 * s1 * c0;
+			_R_to_earth(1, 2) = s0 * s2 - s1 * c0 * c2;
+			_R_to_earth(2, 0) = -s2 * c1;
+			_R_to_earth(2, 1) = s1;
+
+		}
+
+		// calculate the amount that the quaternion has changed by
+		Quatf quat_after_reset = _R_to_earth;
+		Quatf q_error =  quat_after_reset * quat_before_reset.inversed();
+		q_error.normalize();
+
+		// update quaternion states
+		_state.quat_nominal = quat_after_reset;
+		uncorrelateQuatFromOtherStates();
+
+		// record the state change
+		_state_reset_status.quat_change = q_error;
+
+		// update transformation matrix from body to world frame using the current estimate
+		_R_to_earth = Dcmf(_state.quat_nominal);
+
+		// reset the rotation from the EV to EKF frame of reference if it is being used
+		if ((_params.fusion_mode & MASK_ROTATE_EV) && !_control_status.flags.ev_yaw) {
+			calcExtVisRotMat();
+		}
+
+		// update the yaw angle variance using half the nominal yaw separation between models
+		increaseQuatYawErrVariance(sq(fmaxf(M_PI_F / (float)N_MODELS_EKFGSF, 1.0e-2f)));
+
+		// add the reset amount to the output observer buffered data
+		for (uint8_t i = 0; i < _output_buffer.get_length(); i++) {
+			_output_buffer[i].quat_nominal = _state_reset_status.quat_change * _output_buffer[i].quat_nominal;
+		}
+
+		// apply the change in attitude quaternion to our newest quaternion estimate
+		// which was already taken out from the output buffer
+		_output_new.quat_nominal = _state_reset_status.quat_change * _output_new.quat_nominal;
+
+		// capture the reset event
+		_state_reset_status.quat_counter++;
+
+		// reset velocity and position states to GPS - if yaw is fixed then the filter should start to operate correctly
+		resetVelocity();
+		resetPosition();
```

Comment:
In `realignYawGPS` we're also requesting a vel/pos reset but through a flag: https://github.com/PX4/ecl/blob/3fa5f501ae7edc4a793f0973e156706afcc8b419/EKF/ekf_helper.cpp#L463 , why not here?

Reply:
This is at the end of the processing chain for an EKF fusion time horizon update and i was worried about a race condition with other processes that could call a position or velocity timeout. We can experiment with doing that n replay and see if it matters.

Reply:
This code also ensures the reset will happen on this frame rather than waiting for the next GPS message which could be up to 200msec later.